core/remotes/docker/fetcher.go

```diff
@@ -318,19 +347,108 @@ func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string,
 			// Could use buffer pool here but this case should be rare
 			n, err := io.Copy(io.Discard, io.LimitReader(resp.Body, offset))
 			if err != nil {
+				if r.config.Limiter != nil {
+					r.config.Limiter.Release(parallelism)
+				}
 				return nil, fmt.Errorf("failed to discard to offset: %w", err)
 			}
 			if n != offset {
+				if r.config.Limiter != nil {
+					r.config.Limiter.Release(parallelism)
+				}
 				return nil, errors.New("unable to discard to offset")
 			}
 
+			// content range ignored, we can't do concurrent fetches here.
+			if parallelism > 1 && r.config.Limiter != nil {
+				r.config.Limiter.Release(parallelism - 1)
+			}
+			parallelism = 1
 		}
 	}
 
 	body := resp.Body
 	encoding := strings.FieldsFunc(resp.Header.Get("Content-Encoding"), func(r rune) bool {
 		return r == ' ' || r == '\t' || r == ','
 	})
+
+	totalSize, _ := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 0)
+	remaining := totalSize - offset
+	if parallelism > 1 && chunkSize > 0 && remaining > minimumSizeForParallelDL && req.body == nil {
+		// If we have a content length, we can use multiple requests to fetch
+		// the content in parallel. This will make download of bigger bodies
+		// faster, at the cost of parallelism more requests and max
+		// ~(max_parallelism * goroutine footprint) memory usage. The goroutine
+		// footprint should be: the goroutine stack + copy buffer size
+		numChunks := remaining / chunkSize
+		if numChunks*chunkSize < remaining {
+			numChunks++
+		}
+		queue := make(chan int64, parallelism)
+		stopChan := make(chan struct{})
+		readers, writers := make([]io.Reader, numChunks), make([]*io.PipeWriter, numChunks)
+		for i := int64(0); i < numChunks; i++ {
+			readers[i], writers[i] = io.Pipe()
+		}
+		go func() {
+			for i := int64(0); i < numChunks; i++ {
+				queue <- i
+			}
+			close(queue)
+		}()
+		for p := int64(0); p < parallelism; p++ {
+			// start parallel download workers
+			go func() {
+				if r.config.Limiter != nil {
+					defer r.config.Limiter.Release(1)
+				}
+				for i := range queue { // first in first out
+					select {
+					case <-stopChan:
+						return
+					case <-ctx.Done():
+						return
+					default:
+					}
+					if i == 0 {
+						_, err = io.Copy(writers[i], io.LimitReader(resp.Body, chunkSize))
+						_ = resp.Body.Close()
+						_ = writers[i].CloseWithError(err)
+						continue
+					}
+					reqClone := req.clone()
+					reqClone.header.Set("Range", fmt.Sprintf("bytes=%d-", offset+i*chunkSize))
```

Comment:
Thank you for the PR. I was glancing through it out of curiosity (may have more questions later). Why would we set just the start of the range but not the end?

Reply:
Hm, I just never did it, mainly because it felt like we could close the connection when done and avoid having to calculate that. Do you think this would be important ?